backend/src/ee/services/access-approval-request/access-approval-request-service.ts

```diff
@@ -0,0 +1,360 @@
+    const { membership } = await permissionService.getProjectPermission(
+      actor,
+      actorId,
+      project.id,
+      actorAuthMethod,
+      actorOrgId
+    );
+    if (!membership) throw new UnauthorizedError({ message: "You are not a member of this project" });
+
+    const requestedByUser = await userDAL.findUserByProjectMembershipId(membership.id);
+    if (!requestedByUser) throw new UnauthorizedError({ message: "User not found" });
+
+    await projectDAL.checkProjectUpgradeStatus(project.id);
+
+    const { envSlug, secretPath, accessTypes } = verifyRequestedPermissions({ permissions: requestedPermissions });
+    const environment = await projectEnvDAL.findOne({ projectId: project.id, slug: envSlug });
+
+    if (!environment) throw new UnauthorizedError({ message: "Environment not found" });
+
+    const policy = await accessApprovalPolicyDAL.findOne({
+      envId: environment.id,
+      secretPath
+    });
+    if (!policy) throw new UnauthorizedError({ message: "No policy matching criteria was found." });
+
+    const approvers = await accessApprovalPolicyApproverDAL.find({
+      policyId: policy.id
+    });
+
+    const approverUsers = await userDAL.findUsersByProjectMembershipIds(
+      approvers.map((approver) => approver.approverId)
+    );
+
+    if (approverUsers.length !== approvers.length) {
+      throw new BadRequestError({ message: "Some approvers were not found" });
+    }
+
+    const duplicateRequest = await accessApprovalRequestDAL.findOne({
+      policyId: policy.id,
+      requestedBy: membership.id,
+      permissions: JSON.stringify(requestedPermissions),
+      isTemporary
+    });
+
+    if (duplicateRequest) {
+      if (duplicateRequest.privilegeId) {
+        const privilege = await additionalPrivilegeDAL.findById(duplicateRequest.privilegeId);
+
+        const isExpired = new Date() > new Date(privilege.temporaryAccessEndTime || ("" as string));
+
+        if (!isExpired || !privilege.isTemporary) {
+          throw new BadRequestError({ message: "You already have an active privilege with the same criteria" });
+        }
+      } else {
+        const reviewers = await accessApprovalRequestReviewerDAL.find({
+          requestId: duplicateRequest.id
+        });
+
+        const isRejected = reviewers.some((reviewer) => reviewer.status === ApprovalStatus.REJECTED);
+
+        if (!isRejected) {
+          throw new BadRequestError({ message: "You already have a pending access request with the same criteria" });
+        }
+      }
+    }
+
+    const approval = await accessApprovalRequestDAL.transaction(async (tx) => {
+      const approvalRequest = await accessApprovalRequestDAL.create(
+        {
+          policyId: policy.id,
+          requestedBy: membership.id,
+          temporaryRange: temporaryRange || null,
+          permissions: JSON.stringify(requestedPermissions),
+          isTemporary
+        },
+        tx
+      );
+
+      await smtpService.sendMail({
```

Comment:
This doesn't have to be in the transaction — We don't need transactions for this function imo.

You can create the access approval request without the transaction. Then, if it fails, it will throw an error anyways which will mean that it won't ever reach `.sendMail()`.

Reply:
This should be a transaction I'd say. If sendMail fails, something underlying has gone wrong, and the request shouldn't be processed I'd say. It's a safeguard that I wouldn't mind being in place. 

Note: sendMail **won't** fail if it's self-hosted, and the user hasn't set up SMTP.